pkg/operator/csidriveroperator/crcontroller.go

```diff
@@ -0,0 +1,266 @@
+	err := resourceapply.SetSpecHashAnnotation(&requiredOriginal.ObjectMeta, requiredOriginal.Spec)
+	if err != nil {
+		return nil, false, err
+	}
+
+	required := requiredOriginal.DeepCopy()
+	if required.Annotations == nil {
+		required.Annotations = map[string]string{}
+	}
+
+	existing, err := c.operatorClientSet.OperatorV1().ClusterCSIDrivers().Get(context.TODO(), required.Name, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		actual, err := c.operatorClientSet.OperatorV1().ClusterCSIDrivers().Create(context.TODO(), required, metav1.CreateOptions{})
+		reportCreateEvent(c.eventRecorder, required, err)
+		return actual, true, err
+	}
+	if err != nil {
+		return nil, false, err
+	}
+
+	modified := resourcemerge.BoolPtr(false)
+	existingCopy := existing.DeepCopy()
+
+	resourcemerge.EnsureObjectMeta(modified, &existingCopy.ObjectMeta, required.ObjectMeta)
+	// there was no change to metadata, the generation was right, and we weren't asked for force the deployment
+	if !*modified && existingCopy.ObjectMeta.Generation == expectedGeneration {
+		return existingCopy, false, nil
+	}
+
+	// at this point we know that we're going to perform a write.  We're just trying to get the object correct
+	toWrite := existingCopy // shallow copy so the code reads easier
+	toWrite.Spec = *required.Spec.DeepCopy()
+	if klog.V(4).Enabled() {
+		klog.Infof("ClusterCSIDriver %q changes: %v", required.Name, resourceapply.JSONPatchNoError(existing, toWrite))
+	}
+
+	actual, err := c.operatorClientSet.OperatorV1().ClusterCSIDrivers().Update(context.TODO(), toWrite, metav1.UpdateOptions{})
+	reportUpdateEvent(c.eventRecorder, required, err)
+	return actual, true, err
+}
+
+func reportCreateEvent(recorder events.Recorder, obj runtime.Object, originalErr error) {
+	gvk := resourcehelper.GuessObjectGroupVersionKind(obj)
+	if originalErr == nil {
+		recorder.Eventf(fmt.Sprintf("%sCreated", gvk.Kind), "Created %s because it was missing", resourcehelper.FormatResourceForCLIWithNamespace(obj))
+		return
+	}
+	recorder.Warningf(fmt.Sprintf("%sCreateFailed", gvk.Kind), "Failed to create %s: %v", resourcehelper.FormatResourceForCLIWithNamespace(obj), originalErr)
+}
+
+func reportUpdateEvent(recorder events.Recorder, obj runtime.Object, originalErr error, details ...string) {
+	gvk := resourcehelper.GuessObjectGroupVersionKind(obj)
+	switch {
+	case originalErr != nil:
+		recorder.Warningf(fmt.Sprintf("%sUpdateFailed", gvk.Kind), "Failed to update %s: %v", resourcehelper.FormatResourceForCLIWithNamespace(obj), originalErr)
+	case len(details) == 0:
+		recorder.Eventf(fmt.Sprintf("%sUpdated", gvk.Kind), "Updated %s because it changed", resourcehelper.FormatResourceForCLIWithNamespace(obj))
+	default:
+		recorder.Eventf(fmt.Sprintf("%sUpdated", gvk.Kind), "Updated %s:\n%s", resourcehelper.FormatResourceForCLIWithNamespace(obj), strings.Join(details, "\n"))
+	}
+}
```

Comment:
nit: these `report*Event()` are generic enough to live in `util.go`

Reply:
Moved (and squashed)